[FRONTEND] Add the signup form component

ozdenizdolu opened the pull request:
## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Description

Added `SignUpForm.js` to the folder `./src/components/`. 

This component renders a standard signup form that is compatible with our design. The component has only one prop which is `onSubmit`. The component sends all the data it received to the function given to `onSubmit` prop as a single properly labeled JSON object.

The component checks the validity of the input but the validity checks that are being done is insufficient.

The component has a minimum height constraint because of the input boxes but it is pretty flexible on the width and it is going to fill the container it is placed in.

# Screenshots

- **The signup form**

<img alt="Example 1" src="https://user-images.githubusercontent.com/32651175/99113441-500fcb80-2600-11eb-9ec3-6d3d6ffe6052.jpg" width="45%" /><img alt="Example 2" src="https://user-images.githubusercontent.com/32651175/99113490-64ec5f00-2600-11eb-9a7e-51c3b5fb45b1.jpg" width="45%" />

- **JSON data that the function onSubmit is going to receive upon form submission**

<img alt="The data form sends to the onSubmit function" src="https://user-images.githubusercontent.com/32651175/99113358-2bb3ef00-2600-11eb-97cd-6e89576cc416.jpg" />

# Test results

No testing has done other than by hand. It provides basic functionality in non-extreme cases.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have checked my code and corrected any misspellings
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Potential problems/Important remarks

- **Get Captcha** button is non-functional. 
- It might be helpful to implement a function to initialize the values of the form from a given data.

Commit:
Create a branch for making signup form

Commit:
implemented the SignUpForm.js

Commit:
Merge branch 'master' into F-signup-form

MehdiSaffar commented on ecommerce/frontend/src/components/SignUpForm.js:
```diff
@@ -0,0 +1,289 @@
+            <Form.Item
+                name={['address', 'state']}
+                label="State"
+                rules={[
+                    {
+                        required: true,
+                        message: 'Please select your state!',
+                    },
+                ]}
+            >
+                <Input />
+            </Form.Item>
+
+            {/* Zip code input */}
+            <Form.Item
+                name={['address', 'zipcode']}
+                label="Zip Code"
+                rules={[
+                    {
+                        required: true,
+                        message: 'Please enter your zip code!',
+                    },
+                ]}
+            >
+                <Input />
+            </Form.Item>
+            
+            {/* Main address input */}
+            <Form.Item
+                name={['address', 'mainaddress']}
+                label="Main Address"
+                rules={[
+                    {
+                        required: true,
+                        message: 'Please enter your address!',
+                    },
+                ]}
+            >
+                <Input />
+            </Form.Item>
```
1. for keys please use camelCase (example mainaddress -> mainAddress, zipcode -> zipCode). It is what we use in JS.
2.

**1.1.1.2.6** Location (not mandatory for customer and admin users)

The location information is not mandatory for customer users. Therefore the `required: true` needs to depend on whether the Guest user is registering as a Customer or Vendor. This might difficult since required needs to react to value change of a form item so I can help you with it when you are available.

ozdenizdolu replied:
1. I generally automatically use camelCase but I missed those, ill correct it. 
2. Maybe we can set a `onChange` callback on the radio buttons and define a state like `locationEnabled` and make it both disable the input boxes and set `required: locationEnabled`. Will that work?

ozdenizdolu replied:
corrected the non camelCase stuff on 4th commit

MehdiSaffar replied:
for 2:

That sounds like it could work actually. Can you try it? If it still doesn't work then let me know.

MehdiSaffar commented on ecommerce/frontend/src/components/SignUpForm.js:
```diff
@@ -0,0 +1,289 @@
+                </Input.Group>
+            </Form.Item>
+
+            {/* Captcha input
+            TODO: This input is non-functional */}
+            <Form.Item label="Captcha" extra="We must make sure that your are a human.">
+                <Row gutter={8}>
+                    <Col span={12}>
+                        <Form.Item
+                            name="captcha"
+                            noStyle
+                            rules={[
+                                {
+                                    required: true,
+                                    message: 'Please input the captcha you got!',
+                                },
+                            ]}
+                        >
+                            <Input />
+                        </Form.Item>
+                    </Col>
+                    <Col span={12}>
+                        <Button>Get captcha</Button>
+                    </Col>
+                </Row>
+            </Form.Item>
```
Good thinking. Not sure how we will do it but maybe using this library would help https://www.npmjs.com/package/react-google-recaptcha

ozdenizdolu replied:
Let me try to implement this, I'll ask help if I have a hard time

MehdiSaffar commented on ecommerce/frontend/src/components/SignUpForm.js:
```diff
@@ -0,0 +1,289 @@
+                ]}
+                {...tailFormItemLayout}
+            >
+                <Checkbox>
+                    I have read the <a href="">agreement</a>
+                </Checkbox>
```
Note to self: need to either ask customer for some agreement document or just generate a generic one.

MehdiSaffar reviewed (changes requested):
Overall I like it very much!

Like I said, please have the keys in camelCase and try to see if you can integrate ReCAPTCHA. If you need help we can pair code together. If ReCAPTCHA takes too long we can leave it for another PR.

Commit:
correction on some names (made them camelCase)

Commit:
Merge remote-tracking branch 'origin/master' into F-signup-form

Commit:
debugging purposes

abdullahyldz commented:
Thanks for the effort, Ozdeniz. Looking good!
I used your component by adding  <SignUpForm></SignUpForm> in the SignupPage.js Is it the correct way of using it?
When I do so, your component and footer css are mixed to each other. 
![image](https://user-images.githubusercontent.com/44057640/99146951-c4dd1700-268d-11eb-85a1-d45d61cde036.png)
But, when I zoom out, they look ok. 

Also, when I press "Get Captcha" button, my browser page shakes :) This happens only 75% zoom. 

I think, both issues are related to global css.

abdullahyldz commented on ecommerce/frontend/src/components/SignUpForm.js:
```diff
@@ -0,0 +1,311 @@
+                label="Zip Code"
+                rules={[
+                    {
+                        required: addressEnabled,
```
ZipCode Rules: Is there a rule for 'only numeric' input? 
Phone number Rules: Is there a rule for 'certain lengths'? 

Maybe we can improve the form answers.

ozdenizdolu replied:
Mehdi and I worked on that today, Mehdi pushed a corrected version. So they will be controlled.

ozdenizdolu replied:
I think for your previous comment(about the page is being messy and footer floating on top of another): it feels like it is related to the layout. I believe we should tackle that problem when putting everything together.

MehdiSaffar committed:
Signup form enhancements and bug fixes

- Give initial value to register as customer radio button
- Fix error when switching from customer to vendor
- Add simple recaptcha
- Add validation to phone country code and number
- Make phone number an InputNumber
- Correctly hide (but preserve) input fields of Vendor only

Commit:
Changes that have been made:
- Zip code can only be a number
- User agreement Modal pops up when clicked. 
- Minor corrections on typos.

ozdenizdolu commented:
Okay guys, thanks for the reviews. I think I have finalized it. If you have time you can check for one last time if you want. The form looks like as follows right now:
![1](https://user-images.githubusercontent.com/32651175/99155998-5b312d00-26ce-11eb-8c23-1811427300e1.jpg)
![2](https://user-images.githubusercontent.com/32651175/99156001-5f5d4a80-26ce-11eb-9f4e-efb969b90981.jpg)
![3](https://user-images.githubusercontent.com/32651175/99156002-61270e00-26ce-11eb-925a-d9d06b770b29.jpg)

MehdiSaffar commented on ecommerce/frontend/src/components/SignUpForm.js:
```diff
@@ -131,159 +167,187 @@ export const SignUpForm = ({ onSubmit = () => { } }) => {
+                                required: true,
+                                message: 'Please enter your country code!',
+                            },
+                            {
+                                message: 'Wrong country code format',
+                                validator: (rule, val) =>
+                                    /\+\d{1,3}/.test(val) ? Promise.resolve(true) : Promise.reject("Country code format is incorrect!"),
+                            },
```
nitpick: if you have time, make it so that the validator returns an error only when both the field is not empty and the value given does not fit the format. this is done so that you dont get both "please enter country code" and "wrong country code format" when the field is empty. this suggestion is optional, we might leave it for another PR. if you are going to do it, do it for all the similar fields.

MehdiSaffar commented on ecommerce/frontend/src/components/SignUpForm.js:
```diff
@@ -131,159 +167,187 @@ export const SignUpForm = ({ onSubmit = () => { } }) => {
+                <Modal
+                    visible={userAgreementVisible}
+                    title="User Agreement"
+                    onOk={onModalAgreementAccept}
+                    onCancel={onModalAgreementReject}
+                    bodyStyle={{height: 500}}
+                    footer={[
+                <Button key="back" type="primary" onClick={onModalAgreementReject}>
+                    I reject
+                </Button>,
+                <Button key="submit" type="primary" onClick={onModalAgreementAccept}>
+                    I accept
+                </Button>,
```
not sure if you need the callbacks both on onOK and on the button itself. It might make the callback be called twice.

ozdenizdolu replied:
I have checked for double callback case. It only calls them once. So I am not going to change anything here.

MehdiSaffar commented on ecommerce/frontend/src/index.js:
```diff
@@ -3,11 +3,12 @@ import ReactDOM from 'react-dom'
 ReactDOM.render(
-    <React.StrictMode>
-        <App />
-    </React.StrictMode>,
+    // <React.StrictMode>
+    <SignUpForm />,
+    // </React.StrictMode>,
     document.getElementById('root'),
 )
```
Replace SignUpForm with App :D

ozdenizdolu replied:
Will do.

MehdiSaffar reviewed (approved):
Overall very good job, thank you for your effort.

I gave a few nitpicks, you are free to do them if you have time.

You can squash and merge to master.

abdullahyldz reviewed (approved):
Looking good.

Commit:
Finalizing the pull request reviews:
- Doesn't show double error prompt when a field is left blank.
- <App /> added back to the index.js

ozdenizdolu commented:
I am merging the pull request. Final form of the forms looks like as follows: 
![1](https://user-images.githubusercontent.com/32651175/99397717-7a110880-28f4-11eb-9b2a-54684568d1f2.jpg)
![2](https://user-images.githubusercontent.com/32651175/99397720-7b423580-28f4-11eb-8525-f6c4d7ef489c.jpg)
